types/proposal.go

```diff
@@ -84,15 +84,20 @@ func (p *Proposal) ValidateBasic() error {
 // configured Precision and MsgDelay parameters.
 // Specifically, a proposed block timestamp is considered timely if it is satisfies the following inequalities:
 //
-// proposedBlockTime > validatorLocaltime - Precision && proposedBlockTime < validatorLocalTime + Precision + MsgDelay.
+// localtime >= proposedBlockTime - Precision
+// localtime <= proposedBlockTime + MsgDelay + Precision
 //
 // For more information on the meaning of 'timely', see the proposer-based timestamp specification:
 // https://github.com/tendermint/spec/tree/master/spec/consensus/proposer-based-timestamp
-func (p *Proposal) IsTimely(clock tmtime.Source, tp TimingParams) bool {
-	lt := clock.Now()
-	lhs := lt.Add(-tp.Precision)
-	rhs := lt.Add(tp.Precision).Add(tp.MessageDelay)
-	if lhs.Before(p.Timestamp) && rhs.After(p.Timestamp) {
+func (p *Proposal) IsTimely(recvTime time.Time, tp TimingParams, initialHeight int64) bool {
+	// lhs is `proposedBlockTime - Precision` in the first inequality
+	lhs := p.Timestamp.Add(-tp.Precision)
+	// rhs is `proposedBlockTime + MsgDelay + Precision` in the second inequality
+	rhs := p.Timestamp.Add(tp.MessageDelay).Add(tp.Precision)
+
+	recvTimeAfterOrEqLHS := recvTime.After(lhs) || recvTime.Equal(lhs)
+	recvTimeBeforeOrEqRHS := recvTime.Before(rhs) || recvTime.Equal(rhs)
+	if recvTimeAfterOrEqLHS && (p.Height == initialHeight || recvTimeBeforeOrEqRHS) {
 		return true
 	}
 	return false
```

Comment:
Should have a log message here.

Reply:
I'm not usually in favor of logging in helper methods since the helper may get used for many reasons and doesn't necessarily indicate that anything specific is actually happening. Perhaps log by the call sites?

Reply:
There is a log in the caller but if we want to show which of the `false` case occurs and why (i.e. with timestamp and timing values) we pretty much repeat the helper logic in this case.

Reply:
added more data in the log.